### PR TITLE
HAAR-862 amend user email

### PR DESCRIPTION
**Pull request by AlTurner-MOJ**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/auth/model/AuthService.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package uk.gov.justice.digital.hmpps.manageusersapi.adapter.auth.model
+
+data class AuthService(
+  val code: String,
+  val name: String,
+  val description: String?,
+  val contact: String?,
+  val url: String?,
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/auth/model/AuthService.kt:8]: I think url is not nullable in the db

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/external/UserController.kt [hunk 2]
```diff
@@ -155,14 +160,94 @@ class UserController(
...
+      userId,
+      amendUser.email,
+      setPasswordUrl
+    )
+    return if (smokeTestEnabled) resetLink else null
+  }
+
+  private fun createInitialPasswordUrl(@Parameter(hidden = true) request: HttpServletRequest): String {
+    val requestURL = request.requestURL
+    return requestURL.toString().replaceFirst("/api/authuser/.*".toRegex(), "/initial-password?token=")
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/external/UserController.kt:242]: I don't think is quite right as the url is no longer /api/authuser

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailService.kt [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      "firstName" to userDetails.firstName,
+      "fullName" to "${userDetails.firstName} ${userDetails.lastName}",
+      "verifyLink" to verifyLink
+    )
+
+    notificationService.send(notifyTemplateId, parameters, "VerifyEmailRequest", NotificationDetails(userDetails.username, email!!))
+    return LinkEmailAndUsername(verifyLink, email, usernameToUpdate)
+  }
+
+  fun confirmUsernameForUpdate(newEmail: String, existingUsername: String): String {
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailService.kt:56]: very minor - maybe rename to confirmUsernameValidForUpdate

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/auth/model/AuthService.kt (hunk 1 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/auth/model/AuthService.kt]: Amended

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/UserService.kt [hunk 4]
```diff
@@ -30,7 +53,75 @@ class UserService(
...
-    userApiService.disableUserById(userId, deactivateReason)
+    externalUsersApiService.disableUserById(userId, deactivateReason)
+
+  fun amendUserEmailByUserId(
+    userId: UUID,
+    emailAddressInput: String?,
+    url: String,
+  ): String {
+
+    val user = externalUsersSearchApiService.findByUserId(userId)
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/UserService.kt:64]: Should this now be using the external users service client to determine the userId (and the external users endpoint having a role on it?)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/external/UserApiService.kt [hunk 5]
```diff
@@ -17,6 +17,11 @@ class UserApiService(
     val log: Logger = LoggerFactory.getLogger(this::class.java)
   }
 
+  fun updateUserEmailAddressAndUsername(userId: UUID, username: String, email: String) =
+    externalUsersWebClientUtils.put("/users/id/$userId/email", mapOf("username" to username, "email" to email))
+
+  fun hasPassword(userId: UUID) = externalUsersWebClientUtils.get("/users/$userId/password/present", Boolean::class.java)
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/external/UserApiService.kt:23]: You will need my latest commit for this to work correctly - I think the external users endpoint for this should also have a role

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailServiceTest.kt [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+    private fun verifyPrimaryEmailFailure(email: String, reason: String) {
+      assertThatThrownBy { verifyEmailService.validateEmailAddress(email, EmailType.PRIMARY) }.isInstanceOf(
+        VerifyEmailService.ValidEmailException::class.java
+      ).extracting("reason").isEqualTo(reason)
+    }
+  }
+
+  private fun createSampleUserWith(username: String, email: String): ExternalUserDetailsDto {
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailServiceTest.kt:226]: Should prob pull this out into a Helper class for everybody to use for testing

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailServiceTest.kt [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+      verifyPrimaryEmailFailure("a@b.&com", "characters")
+    }
+
+    @Test
+    fun invalidDomain() {
+      verifyPrimaryEmailFailure("a@b.com", "domain")
+      verify(verifyEmailDomainService).isValidEmailDomain("b.com")
+    }
+
+    @Test
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailServiceTest.kt:207]: Do we need a test for secondary - just to show it doesn't do anything?  I'm not totally sure if we do or not

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/external/UserController.kt (hunk 2 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/external/UserController.kt]: Corrected - now uses `api.base.url.oauth` environment variable

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailService.kt (hunk 3 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailService.kt]: Done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailServiceTest.kt (hunk 7 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailServiceTest.kt:207]: Not yet I don't think - I should probably remove it from the enum for now as it is not used?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/UserService.kt (hunk 4 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/UserService.kt:64]: Not sure, but could we pick this up under a separate ticket if it is going to need changes to external-users-api too?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/external/UserApiService.kt (hunk 5 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/external/UserApiService.kt]: See comment above?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/UserService.kt [hunk 8]
```diff
@@ -3,23 +3,46 @@ package uk.gov.justice.digital.hmpps.manageusersapi.service.external
...
-  private val userApiService: UserApiService,
-  private val emailNotificationService: EmailNotificationService,
+  private val notificationService: NotificationService,
+  private val externalUsersApiService: UserApiService,
+  private val externalUsersSearchApiService: UserSearchApiService,
+  private val authApiService: AuthApiService,
+  private val userGroupApiService: UserGroupApiService,
+  private val verifyEmailService: VerifyEmailService,
   private val telemetryClient: TelemetryClient,
+  @Value("\${api.base.url.oauth}") private val authBaseUri: String,
```
[the remark this answers is not in the document]
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/UserService.kt:28]: OK, happy to wait if you want to commit ahead of this?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/UserService.kt (hunk 4 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/UserService.kt:64]: Discussed with @karenmillermoj - both end points have same roles therefore we agreed OK

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/external/UserApiService.kt (hunk 5 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/external/UserApiService.kt]: Also discussed with @karenmillermoj and verified permissions in external-users-api all align with user permissions on original Auth end point, so OK

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailServiceTest.kt (hunk 6 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailServiceTest.kt]: Done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailServiceTest.kt (hunk 7 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailServiceTest.kt:207]: Removed SECONDARY from enum, temporarily

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/fixtures/UserFixture.kt [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+import java.util.UUID
+import uk.gov.justice.digital.hmpps.manageusersapi.resource.external.ExternalUserDetailsDto
+
+class UserFixture {
+
+    fun createExternalUserDetails(userId: UUID, username: String, email: String): ExternalUserDetailsDto {
+        return ExternalUserDetailsDto(
+            userId = userId,
+            username = username,
+            email = email,
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/fixtures/UserFixture.kt]: I would prob put defaults on these and make the fields optional.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/fixtures/UserFixture.kt [hunk 10]
```diff
@@ -0,0 +1,23 @@
+package uk.gov.justice.digital.hmpps.manageusersapi.fixtures
+
+import java.time.LocalDateTime
+import java.util.UUID
+import uk.gov.justice.digital.hmpps.manageusersapi.resource.external.ExternalUserDetailsDto
+
+class UserFixture {
+
+    fun createExternalUserDetails(userId: UUID, username: String, email: String): ExternalUserDetailsDto {
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/fixtures/UserFixture.kt]: If you make a companion object with createExternalUserDetails in it - then you can easily import the UserFixture.Companion. createExternalUserDetails in the Tests

**Review by karenmillermoj — DISMISSED** [review]
lgtm

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/fixtures/UserFixture.kt (hunk 9 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/fixtures/UserFixture.kt]: Done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/fixtures/UserFixture.kt (hunk 10 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/fixtures/UserFixture.kt]: Done

**Pull request merged**